### ISSUE TITLE
[BUG] SofT-GRPO training freezes with ResourceWarning (unclosed file) and Ray GCS crashes on A100 40GB (both single and dual GPU)

### ISSUE THREAD

**Issue by xiaocen-luo** [zz1358m/SofT-GRPO-master#5]
Hi @softgrpo-team,

I'm trying to train Llama-3.2-1B-Instruct using the SofT-GRPO method with the verl-0.4.x framework. I've tested on both **single A100 40GB** and **dual A100 40GB** configurations, and encountered similar persistent issues. I would really appreciate your guidance.

### My Setup:
- **Model**: Llama-3.2-1B-Instruct
- **GPU**: Tested on both single A100 40GB and dual A100 40GB
- **Framework**: verl-0.4.x with SofT-GRPO implementation
- **Dataset**: DAPO math dataset (train-dapo-verl.parquet)

### Key Training Configuration:
data.train_batch_size=1 (single GPU) / 2 (dual GPU)
data.max_prompt_length=512
data.max_response_length=1024
actor_rollout_ref.rollout.gpu_memory_utilization=0.2 (single) / 0.3 (dual)
actor_rollout_ref.rollout.enable_soft_thinking=True
actor_rollout_ref.rollout.free_cache_engine=True
actor_rollout_ref.rollout.enable_chunked_prefill=False

### Issues Observed (on both single and dual GPU setups):

1. **Ray GCS Service Crashes**: After running ~27-240 training steps, the Ray GCS service crashes with the following error:
GCS consider this node to be dead. This may happen when GCS is not backed by a DB and restarted or there is data loss in the DB.
Raylet is terminated. Termination is unexpected. Possible reasons include: (1) SIGKILL by the user or system OOM killer, (2) Invalid memory access from Raylet causing SIGSEGV or SIGBUS, (3) Other termination signals.

2. **Program Freezes with ResourceWarning**: The training often hangs/freezes with the following warnings appearing repeatedly:
ResourceWarning: Implicitly cleaning up <TemporaryDirectory '/tmp/tmp...'>
ResourceWarning: unclosed file <_io.BufferedWriter name=21>
After these warnings appear, the program becomes unresponsive and cannot be interrupted with Ctrl+C. We have to manually kill the processes or restart the node. This happens on both single and dual GPU setups.

3. **Memory Issues**: 
- GPU memory usage: ~25.3GB allocated, ~29.8GB reserved (on a 40GB card)
- CPU memory usage: ~43-55GB during training
- The memory usage seems to grow gradually over time, suggesting a potential memory leak

4. **Training Progress**: The training runs for a while (27-240 steps) but then either freezes or crashes. The model appears to be learning (pg_loss ~1.0, grad_norm > 20), but training cannot complete a full epoch.

### What I've Tried:
- Tested on both single and dual GPU setups (both show similar issues)
- Lowering `gpu_memory_utilization` from 0.6 → 0.3 → 0.25 → 0.2 → 0.15
- Setting `free_cache_engine=True`
- Setting `enable_chunked_prefill=False`
- Disabling validation (`val_before_train=False`, `test_freq=0`)
- Using `rollout.load_format=dummy_hf`
- Various Ray GCS timeout configurations:

RAY_gcs_rpc_server_reconnect_timeout_s=60-3600
RAY_gcs_connect_timeout_seconds=60-3600
RAY_health_check_timeout_ms=1200-3600000

- Unsetting `ROCR_VISIBLE_DEVICES` and `HIP_VISIBLE_DEVICES`
- Disabling tensorboard logging (using only console logger)

### Questions:
1. Is this memory usage expected for Llama-3.2-1B with SofT-GRPO? Should I use a larger GPU (e.g., A100 80GB) or can this be optimized further?
2. The `ResourceWarning: unclosed file` seems to come from sglang's tempfile usage. Is this a known issue? The program freezes after these warnings appear. Is there a workaround or fix?
3. Is there a known issue with Ray GCS crashing during long-running SofT-GRPO training? Are there recommended Ray configurations for stability?
4. Could there be a memory leak in the sglang rollout or verl checkpoint management? The memory usage gradually increases over time.
5. Are there any recommended configuration changes for training 1B models on single/dual GPU setups?

### Additional Context:
- Running on SLURM cluster (gpu02 node, tested on both single and dual GPU)
- Python 3.11, PyTorch 2.4.0
- sglang is from the soft-thinking branch (Soft-Thinking+noise+loss-main)
- We've tried both `trainer.n_gpus_per_node=1` and `=2` with similar results
- The ResourceWarning appears regardless of whether soft-thinking is enabled

Any insights or suggestions would be greatly appreciated. Happy to provide more logs or details if needed.

Thanks!

**Comment by xiaocen-luo** [comment]
Thanks,After correctly setting the number of GPUs and CPUs, it no longer hangs.
